src/lib/loaders/batchLoader.ts

```diff
@@ -69,26 +111,40 @@ export const batchLoader = ({
   if (!ENABLE_RESOLVER_BATCHING) {
     return singleLoader ? singleLoader : multipleLoader
   }
-  const dl = new DataLoader(keys => {
-    let groupedKeys = groupKeys(keys as any)
-
-    return Promise.all(
-      groupedKeys.map(keys => {
-        if (keys.id.length === 1 && singleLoader) {
-          return singleLoader(keys.id[0])
-        } else {
-          return multipleLoader(keys)
-        }
-      })
-    ).then(data => {
-      const normalizedResults = data.map((queriedGroup, groupIndex) => {
-        return groupedKeys[groupIndex].id.map(
-          id => queriedGroup.find(r => r._id === id) || defaultResult
+  const dl = new DataLoader(
+    (keys: Key[]) => {
+      const normalKeys = normalizeKeys(keys)
+      const [keyGroups, groupedKeys] = groupKeys(normalKeys)
+
+      return Promise.all(
+        groupedKeys.map(compactKeyForRequest).map(keys => {
+          if (keys.id.length === 1 && singleLoader) {
+            return singleLoader(keys.id[0])
+          } else {
+            return multipleLoader(keys)
+          }
+        })
+      ).then((data: Array<GravityResult | GravityResult[]>) => {
+        const normalizedData = data.map(
+          datum => (Array.isArray(datum) ? datum : [datum])
         )
```

Comment:
Remind me, a `singleLoader` is _not_ a loader that only hits endpoints that return single entities [it may return lists], it's simply an endpoint that does not take a list of IDs to fetch, right? And which is why this normalisation can't happen on line 122 instead.

Reply:
My expectation is that a `singleLoader` generally returns single entities, but I guess there's no guarantee of that.  It could potentially happen on 122. That might save some loops. Still, if a list endpoint ended up sending a single result (not as an array) then it would fail. I'm assuming that would never happen, but 🤷‍♂️